drivers/flash/soc_flash_numaker.c

```diff
@@ -0,0 +1,285 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_fmc
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include "flash_priv.h"
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(flash_numaker, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_WRITE_BLOCK_SIZE DT_PROP_OR(SOC_NV_FLASH_NODE, write_block_size, 0x04)
+
+struct flash_numaker_data {
+	FMC_T *fmc;
+	struct k_sem write_lock;
+	uint32_t flash_block_base;
+};
+
+static const struct flash_parameters flash_numaker_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+/* Validate offset and length */
+static bool flash_numaker_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len == 0)) {
+		return false;
+	}
+
+	/* check for max value */
+	if ((offset + len) < 0) {
```

Comment:
This will not work on a principle how casting in C works. All the parameters for the equation will be promoted to the highest unsigned type in equation, which in this case would be `size_t`, so the result of this equation is basically unsigned.
I do not think it is possible to weasel out, this way, from checking `len` and `offset` against allowed range.

Reply:
Yes, it will casting as unsigned int in C.
I changed to use int64_t to cover it in 32 bit Cortex-M platform.
Update at commit  bcec8a3355338c.

Reply:
Rebase to apply  #60551 to resolve warning `API_DEFINE: do not specify a non-Zephyr API for libc`.

Reply:
Fixed at commit 1e4b14fd571